grouper/models/user_password.py

```diff
@@ -0,0 +1,84 @@
+from datetime import datetime
+import hashlib
+import hmac
+import os
+
+from grouper.models.base.model_base import Model
+from sqlalchemy import Column, Integer, String, UniqueConstraint, ForeignKey, DateTime, Text
+from sqlalchemy.orm import relationship
+
+
+def _make_salt():
+    return os.urandom(20).encode("hex")
+
+
+class UserPassword(Model):
+    """Simple password hashes stored to a user for use by thirdparty applications"""
+
+    __tablename__ = "user_passwords"
+
+    id = Column(Integer, primary_key=True)
+
+    user_id = Column(Integer, ForeignKey("users.id"))
+    name = Column(String(length=16), nullable=False)
+
+    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
+    disabled_at = Column(DateTime, default=None, nullable=True)
+
+    _hashed_secret = Column(Text, nullable=False)
+    salt = Column(String(length=128), nullable=False)
+
+    user = relationship("User")
+
+    __table_args__ = (
+        UniqueConstraint("user_id", "name"),
+    )
+
+    @staticmethod
+    def get(session, user, name=None, id=None):
+        """Retrieves a single UserPassword.
+
+        Args:
+            session (Session): Session object
+            user (User): Owner of the password
+            name (str): Name of the password
+            id (int): Primary key of the password
+
+        Returns:
+            UserPassword: UserPassword matching the specified constraints
+
+        """
+        assert name is None or id is None
+
+        if name is not None:
+            return session.query(UserPassword).filter_by(name=name, user=user).scalar()
+        return session.query(UserPassword).filter_by(id=id, user=user).scalar()
+
+    @property
+    def password(self):
+        return self._hashed_secret
+
+    @password.setter
+    def password(self, new_password):
+        self.salt = _make_salt()
+        self._hashed_secret = hashlib.sha512(new_password + self.salt).hexdigest()
+
+    def check_password(self, password_to_check):
```

Comment:
We could be fancy and override `__eq__`, but that's probably just needlessly magical.